### PR TITLE
feat: add rockspec

### PR DESCRIPTION
**Pull request by teto**
running some experiment with luarocks to handle neovim plugins (especially the dependency part) . Opening the PR to avoid someone else having to do it.

### CODE REVIEW

File: gitsigns-scm-1.rockspec [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+dependencies = {
+   'lua >= 5.1, < 5.4',
+   'plenary',
+}
+
+source = {
+   url = 'http://github.com/TimUntersberger/neogit/v' .. _MODREV .. '.zip',
+   dir = 'neogit-' .. _MODREV,
```
Comment (lewis6991) [inline gitsigns-scm-1.rockspec:23]: What's this for?

File: gitsigns-scm-1.rockspec [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+   detailed = [[
+   	git signs
+
+   ]],
+   homepage = 'http://github.com/lewis6991/gitsigns.nvim',
+   license = 'MIT/X11',
+}
+
+dependencies = {
+   'lua >= 5.1, < 5.4',
```
Comment (lewis6991) [inline gitsigns-scm-1.rockspec:17]: Not sure if anything other than 5.1 actually works.

File: gitsigns-scm-1.rockspec [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+   type = 'builtin',
+   modules = {
+	   ["gitsigns"] = "lua/gitsigns.lua",
+	   ["gitsigns.actions"] = "lua/gitsigns/actions.lua",
+	   ["gitsigns.cache"] = "lua/gitsigns/cache.lua",
+	   ["gitsigns.config"] = "lua/gitsigns/config.lua",
+	   ["gitsigns.current_line_blame"] = "lua/gitsigns/current_line_blame.lua",
+	   ["gitsigns.debounce"] = "lua/gitsigns/debounce.lua",
+	   ["gitsigns.debug"] = "lua/gitsigns/debug.lua",
+	   ["gitsigns.diff"] = "lua/gitsigns/diff.lua",
```
Comment (lewis6991) [inline gitsigns-scm-1.rockspec:43]: Note to self: need to delete this file.

File: gitsigns-scm-1.rockspec [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	   -- -- "ftplugin",
+	   -- 'lua'
+   -- }
+}
+
+-- if _MODREV == 'git' then
+--    dependencies[#dependencies + 1] = 'ldoc'
+
+--    source = {
+--       url = 'git://github.com/luaposix/luaposix.git',
```
Comment (lewis6991) [inline gitsigns-scm-1.rockspec:67]: What's this?

**Review by lewis6991 — COMMENTED** [review]
Do you intend for this to get merged or is it purely for experiments?

**Comment by teto** [comment]
one difficulty with the vimscript plugin ecosystem is the lack of a dependency manager, it limited reuse of code and the emergence of better plugins. Lua has a package manager (luarocks) that has standarized lua dependency management around the  rockspec format (I personnally use nix to install lua packages, and it leverages rockspec). I've seen plugins asking users
to setup dependencies like this:
```
use {
  'lewis6991/gitsigns.nvim',
  requires = {
    'nvim-lua/plenary.nvim'
  }
}
```
but this doesn't scale. Users shouldn't have to deal with dependencies, so I think levearging luarocks is an interesting approach. I started experimenting with gitsigns because I use it and it is a good usecase (1 dependency, few files).
Wanted to know what's your stand on this ? If you agree, I may upload a rockspec on luarocks.org . I can transfer ownership/remove the rockspec later ( or use a random name).
To sum up there is no need to merge, this could be  closed too but I wanted to communicate about it to get your feedback or in case others have input.

**Comment by lewis6991** [comment]
Sounds sensible to me.

It might go without saying, but I think versioning will also be needed for something like this to really work. That is why I've began releasing versions of gitsigns every so often even though I don't expect many people are using them. One usecase where this would have been useful would be when plenary released each async implementation (v1 and v2) and users had trouble upgrading gitsigns because they weren't upgrading plenary with it. A dependency manager would have automatically dealt with that.

But sure, go ahead. I'll be happy to support this once something usuable is in place.

**Comment by teto** [comment]
you are absolutely right. It's not in the vim culture but versioning would be helpful need. It has kinda started for the very active plugin nvim-treesitter which provides a compat-0.5 branch. The plenary example is also excellent. The rockspec format allows to set constriants on dependencies (e.g, <= 4, >= 3, == 2.0)

**Comment by teto** [comment]
I've rebased, could you try reopening to see if it updates github UI ?
